In [ ]:
!pip install kaggle_environments
import kaggle_environments
from kaggle_environments import make, evaluate

In [4]:
import numpy as np
import random
import time

# How deep to make the game tree: higher values take longer to run!
N_STEPS = 2

In [5]:
def get_win_percentages(agent1, agent2, n_rounds=10):
    # Use default Connect Four setup
    import numpy as np
    config = {'rows': 6, 'columns': 7, 'inarow': 4}
    # Agent 1 goes first (roughly) half the time          
    outcomes = evaluate("connectx", [agent1, agent2], config, [], n_rounds//2)
    # Agent 2 goes first (roughly) half the time      
    outcomes += [[b,a] for [a,b] in evaluate("connectx", [agent2, agent1], config, [], n_rounds-n_rounds//2)]
    print("Agent 1 Win Percentage:", np.round(outcomes.count([1,-1])/len(outcomes), 2))
    print("Agent 2 Win Percentage:", np.round(outcomes.count([-1,1])/len(outcomes), 2))
    print("Number of Invalid Plays by Agent 1:", outcomes.count([None, 0]))
    print("Number of Invalid Plays by Agent 2:", outcomes.count([0, None]))

In [25]:
#def my_agent(obs, config):
def heuristic(obs, config):

    ################################
    # Imports and helper functions #
    ################################
    
    import numpy as np
    import random
    
    # How deep to make the game tree: higher values take longer to run!
    N_STEPS = 2

    # Gets board at next step if agent drops piece in selected column
    def drop_piece(grid, col, mark, config):
        next_grid = grid.copy()
        for row in range(config.rows-1, -1, -1):
            if next_grid[row][col] == 0:
                break
        next_grid[row][col] = mark
        return next_grid

    # Helper function for get_heuristic: checks if window satisfies heuristic conditions
    def check_window(window, num_discs, piece, config):
        return (window.count(piece) == num_discs and window.count(0) == config.inarow-num_discs)

    # Helper function for get_heuristic: counts number of windows satisfying specified heuristic conditions
    def count_windows(grid, num_discs, piece, config):
        num_windows = 0
        # horizontal
        for row in range(config.rows):
            for col in range(config.columns-(config.inarow-1)):
                window = list(grid[row, col:col+config.inarow])
                if check_window(window, num_discs, piece, config):
                    num_windows += 1
        # vertical
        for row in range(config.rows-(config.inarow-1)):
            for col in range(config.columns):
                window = list(grid[row:row+config.inarow, col])
                if check_window(window, num_discs, piece, config):
                    num_windows += 1
        # positive diagonal
        for row in range(config.rows-(config.inarow-1)):
            for col in range(config.columns-(config.inarow-1)):
                window = list(grid[range(row, row+config.inarow), range(col, col+config.inarow)])
                if check_window(window, num_discs, piece, config):
                    num_windows += 1
        # negative diagonal
        for row in range(config.inarow-1, config.rows):
            for col in range(config.columns-(config.inarow-1)):
                window = list(grid[range(row, row-config.inarow, -1), range(col, col+config.inarow)])
                if check_window(window, num_discs, piece, config):
                    num_windows += 1
        return num_windows
    
    # Helper function for minimax: calculates value of heuristic for grid
    A = 2
    B = 10
    C = -1
    D = -10
    def get_heuristic(grid, mark, config):
        num_threes = count_windows(grid, 3, mark, config) #A
        num_fours = count_windows(grid, 4, mark, config)  #B
        num_threes_opp = count_windows(grid, 3, mark%2+1, config) #C
        num_fours_opp = count_windows(grid, 4, mark%2+1, config)  #D
        score = A*num_threes + C*num_threes_opp + D*num_fours_opp + B*num_fours
        return score

    # Helper function for minimax: checks if agent or opponent has four in a row in the window
    def is_terminal_window(window, config):
        return window.count(1) == config.inarow or window.count(2) == config.inarow

    # Helper function for minimax: checks if game has ended
    def is_terminal_node(grid, config):
        # Check for draw 
        if list(grid[0, :]).count(0) == 0:
            return True
        # Check for win: horizontal, vertical, or diagonal
        # horizontal 
        for row in range(config.rows):
            for col in range(config.columns-(config.inarow-1)):
                window = list(grid[row, col:col+config.inarow])
                if is_terminal_window(window, config):
                    return True
        # vertical
        for row in range(config.rows-(config.inarow-1)):
            for col in range(config.columns):
                window = list(grid[row:row+config.inarow, col])
                if is_terminal_window(window, config):
                    return True
        # positive diagonal
        for row in range(config.rows-(config.inarow-1)):
            for col in range(config.columns-(config.inarow-1)):
                window = list(grid[range(row, row+config.inarow), range(col, col+config.inarow)])
                if is_terminal_window(window, config):
                    return True
        # negative diagonal
        for row in range(config.inarow-1, config.rows):
            for col in range(config.columns-(config.inarow-1)):
                window = list(grid[range(row, row-config.inarow, -1), range(col, col+config.inarow)])
                if is_terminal_window(window, config):
                    return True
        return False

    # Minimax implementation
    def minimax(node, depth, maximizingPlayer, mark, config):
        is_terminal = is_terminal_node(node, config)
        valid_moves = [c for c in range(config.columns) if node[0][c] == 0]
        if depth == 0 or is_terminal:
            return get_heuristic(node, mark, config)
        if maximizingPlayer:
            value = -np.Inf
            for col in valid_moves:
                child = drop_piece(node, col, mark, config)
                value = max(value, minimax(child, depth-1, False, mark, config))
            return value
        else:
            value = np.Inf
            for col in valid_moves:
                child = drop_piece(node, col, mark%2+1, config)
                value = min(value, minimax(child, depth-1, True, mark, config))
            return value
        
    # Uses minimax to calculate value of dropping piece in selected column
    def score_move(grid, col, mark, config, nsteps):
        next_grid = drop_piece(grid, col, mark, config)
        score = minimax(next_grid, nsteps-1, False, mark, config)
        return score
    
    #########################
    # Agent makes selection #
    #########################
    
    # Get list of valid moves
    valid_moves = [c for c in range(config.columns) if obs.board[c] == 0]
    
    # Convert the board to a 2D grid
    grid = np.asarray(obs.board).reshape(config.rows, config.columns)
    
    # Use the heuristic to assign a score to each possible board in the next step
    scores = dict(zip(valid_moves, [score_move(grid, col, obs.mark, config, N_STEPS) for col in valid_moves]))
    
    # Get a list of columns (moves) that maximize the heuristic
    max_cols = [key for key in scores.keys() if scores[key] == max(scores.values())]
    
    # Select at random from the maximizing columns
    return random.choice(max_cols)

In [27]:
start_time = time.time()
n_rounds=75
get_win_percentages(agent1=heuristic, agent2="negamax", n_rounds=n_rounds)
print ("Total time taken: {} seconds (per round: {} seconds)".format(round(time.time() - start_time, 1), 
                                                                     round(time.time() - start_time)/n_rounds,3))

Agent 1 Win Percentage: 0.87
Agent 2 Win Percentage: 0.12
Number of Invalid Plays by Agent 1: 0
Number of Invalid Plays by Agent 2: 0
Total time taken: 705.2 seconds (per round: 9.4 seconds)


#Heuristic vs. Negamax

* Heuris dep2: Win Percentage: 0.71
* Negamax: Win Percentage: 0.25

Total time taken: 270.4 seconds (per round: 2.7 seconds) 
__________________________________

* Heuris dep3: Win Percentage: 0.87
* Negamax: Win Percentage: 0.12

Total time taken: 705.2 seconds (per round: 9.4 seconds)
__________________________________

#Heuristic vs. Heuristic

* Heuris: (2,10,-1,-10) Win Percentage: 0.61
* Tic: (2,4,-1,-8) Win Percentage: 0.39

Total time taken: 22.4 seconds (per round: 0.22 seconds)
__________________________________

* Heuris: (2,10,-1,-10) Win Percentage: 0.47
* Tic: (2,10,-1,-5) Win Percentage: 0.53

Total time taken: 26.7 seconds (per round: 0.27 seconds)
__________________________________

* Heuris: dep1, (2,10,-1,-10) Win Percentage: 0.07
* Tic: dep2, (2,10,-1,-5) Win Percentage: 0.93

Total time taken: 112.4 seconds (per round: 1.12 seconds)
__________________________________

#Trial Runs

* My_Agent Win Percentage: 0.41
* My_Agent Win Percentage: 0.45
* Number of Invalid Plays by My_Agent: 0
* Number of Invalid Plays by My_Agent: 0

Total time taken: 446.3462007045746 seconds

Time taken per round: 4.4634632921218875 seconds
__________________________________

* Random Win Percentage: 0.04
* One_Step Agent Win Percentage: 0.96
* Number of Invalid Plays by Random: 0
* Number of Invalid Plays by One_Step Agent: 0

Total time taken: 16.603688955307007 seconds

Time taken per round: 0.16603797674179077 seconds
__________________________________

* Random Win Percentage: 0.01
* My_Agent Win Percentage: 0.99
* Number of Invalid Plays by Random: 0
* Number of Invalid Plays by My_Agent: 0

Total time taken: 142.17096710205078 seconds

Time taken per round: 1.4217106986045838 seconds
__________________________________

* One_Step Agent Win Percentage: 0.56
* One_Step Agent Win Percentage: 0.44
* Number of Invalid Plays by One_Step Agent: 0
* Number of Invalid Plays by One_Step Agent: 0

Total time taken: 31.378709077835083 seconds

Time taken per round: 0.3137881112098694 seconds

#Write and submit

In [ ]:
if False:    
    import inspect
    import os

    def write_agent_to_file(function, file):
        with open(file, "a" if os.path.exists(file) else "w") as f:
            f.write(inspect.getsource(function))
            print(function, "written to", file)

    write_agent_to_file(my_agent, "submission.py")

### 1) A closer look

The heuristic from the tutorial looks at all groups of four adjacent grid locations on the same row, column, or diagonal and assigns points for each occurrence of the following patterns:

<center>
<img src="https://i.imgur.com/3NvBEGL.png" width=70%><br/>
</center>

Is it really necessary to use so many numbers to define the heuristic?  Consider simplifying it, as in the image below.

<center>
<img src="https://i.imgur.com/grViegG.png" width=70%><br/>
</center>

How would each heuristic score the potential moves in the example below (where, in this case, the agent looks only one step ahead)?  Which heuristic would lead to the agent selecting the better move?

<center>
<img src="https://i.imgur.com/LWPLy7N.png" width=100%><br/>
</center>

### 5) Submit to the competition

Now, it's time to submit an agent to the competition!  Use the next code cell to define an agent.  (You can see an example of how to write a valid agent in **[this notebook](https://www.kaggle.com/alexisbcook/create-a-connectx-agent)**.)

If you decide to use the minimax code from the tutorial, you might like to add [**alpha-beta pruning**](https://en.wikipedia.org/wiki/Alpha%E2%80%93beta_pruning) to decrease the computation time (i.e., get the minimax algorithm to run much faster!).  In this case, "alpha" and "beta" to refer to two values that are maintained while the algorithm is running, that help to identify early stopping conditions.  

Without alpha-beta pruning, minimax evaluates each leaf node.  With alpha-beta pruning, minimax only evaluates nodes that could provide information that affects the agent's choice of action.  Put another way, it identifies nodes that could not possibly affect the final result and avoids evaluating them.

Then, follow these steps to submit your agent to the competition:
1. Begin by clicking on the blue **Save Version** button in the top right corner of the window.  This will generate a pop-up window.  
2. Ensure that the **Save and Run All** option is selected, and then click on the blue **Save** button.
3. This generates a window in the bottom left corner of the notebook.  After it has finished running, click on the number to the right of the **Save Version** button.  This pulls up a list of versions on the right of the screen.  Click on the ellipsis **(...)** to the right of the most recent version, and select **Open in Viewer**.  This brings you into view mode of the same page. You will need to scroll down to get back to these instructions.
4. Click on the **Output** tab on the right of the screen.  Then, click on the blue **Submit** button to submit your results to the leaderboard.

You have now successfully submitted to the competition!

If you want to keep working to improve your performance, select the blue **Edit** button in the top right of the screen. Then you can change your code and repeat the process. There's a lot of room to improve, and you will climb up the leaderboard as you work.


Go to **"My Submissions"** to view your score and episodes being played.

# Keep going

Move on to learn how to **[use deep reinforcement learning](https://www.kaggle.com/alexisbcook/deep-reinforcement-learning)** to develop an agent without a heuristic!